exchanges/exchange.go

```diff
@@ -309,68 +309,56 @@ func (b *Base) SetCurrencyPairFormat() error {
 	exchangeAssets := b.CurrencyPairs.GetAssetTypes(false)
-	for x := range assetTypes {
-		if !exchangeAssets.Contains(assetTypes[x]) {
-			log.Warnf(log.ExchangeSys,
-				"%s exchange asset type %s unsupported, please manually remove from configuration",
-				b.Name,
-				assetTypes[x])
-			continue // If there are unsupported assets contained in config, skip.
-		}
-
-		var enabledAsset bool
-		if b.Config.CurrencyPairs.IsAssetEnabled(assetTypes[x]) == nil {
-			enabledAsset = true
+	for _, a := range assetTypes {
+		if !exchangeAssets.Contains(a) {
+			log.Warnf(log.ExchangeSys, "%s exchange asset type %s unsupported, please manually remove from configuration", b.Name, a)
+			continue
 		}
-
-		err := b.CurrencyPairs.SetAssetEnabled(assetTypes[x], enabledAsset)
-		// Suppress error when assets are enabled by default and they are being
-		// enabled by config. A check for the inverse
-		// e.g. currency.ErrAssetAlreadyDisabled is not needed.
-		if err != nil && !errors.Is(err, currency.ErrAssetAlreadyEnabled) {
-			return err
+		if !b.Config.CurrencyPairs.UseGlobalFormat {
+			// TODO: Should probably be in SetCurrencyPairFormat. See #1748
+			if err := b.setConfigPairFormatFromExchange(a); err != nil {
+				return err
+			}
 		}
 
-		cfgPS, err := b.Config.CurrencyPairs.Get(assetTypes[x])
+		cfgPS, err := b.Config.CurrencyPairs.Get(a)
 		if err != nil {
 			return err
 		}
-
-		if b.Config.CurrencyPairs.UseGlobalFormat {
-			err = b.CurrencyPairs.StorePairs(assetTypes[x], cfgPS.Available, false)
-			if err != nil {
-				return err
-			}
-			err = b.CurrencyPairs.StorePairs(assetTypes[x], cfgPS.Enabled, true)
-			if err != nil {
-				return err
-			}
-			continue
+		if err := b.CurrencyPairs.StorePairs(a, cfgPS.Available, false); err != nil {
+			return err
 		}
-		exchPS, err := b.CurrencyPairs.Get(assetTypes[x])
-		if err != nil {
+		if err := b.CurrencyPairs.StorePairs(a, cfgPS.Enabled, true); err != nil {
 			return err
 		}
 
-		if exchPS.ConfigFormat != nil {
-			err = b.Config.CurrencyPairs.StoreFormat(assetTypes[x], exchPS.ConfigFormat, true)
-			if err != nil {
-				return err
-			}
+		var enabledAsset bool
+		if b.Config.CurrencyPairs.IsAssetEnabled(a) == nil {
+			enabledAsset = true
 		}
-		if exchPS.RequestFormat != nil {
-			err = b.Config.CurrencyPairs.StoreFormat(assetTypes[x], exchPS.RequestFormat, false)
-			if err != nil {
-				return err
-			}
+
+		// Must happen after StorePairs, which would have enabled the asset automatically
+		if err := b.CurrencyPairs.SetAssetEnabled(a, enabledAsset); err != nil {
+			return err
 		}
+	}
+	return nil
+}
 
-		err = b.CurrencyPairs.StorePairs(assetTypes[x], cfgPS.Available, false)
-		if err != nil {
+// setConfigPairFormatFromExchange sets the config formats from the exchange's format
+// This deprecated behaviour will be removed, because config should not be backloaded from runtime
+func (b *Base) setConfigPairFormatFromExchange(a asset.Item) error {
```

Comment:
Add test please 🌈 🌞

Reply:
Impossible to get to those nested err situations, both in tests, and in runtime, because we're detecting the nils already, and bad asset or nil-guarding is all that StoreFormat would error on.
Nevermind. I've done the best I can.
Definitely don't want to re-inline this.

Fixed https://github.com/gbjk/gocryptotrader/commit/3bf8af59f3

Reply:
Thank you! 🌞